roachtest: assert on CPU in rebalance/by-load

kvoli opened the pull request:
The `rebalance/by-load/replicas` roachtests periodically flake due to a
known limitation, where a cluster with a small number of ranges may not
be properly balanced due to heterogeneous localities (including
multi-store) https://github.com/cockroachdb/cockroach/issues/88829.

This PR updates the total number of ranges from 1 per-store, to 5
per-store for `rebalance/by-load/*` roachtests.

The roachtest asserted on the lease count, as a proxy for load, assuming
that the load evenly hits each lease for the created ranges. However,
the principal indicator of load in a read heavy workload is CPU.

This PR updates the test assertion to require that every store's CPU
is within 10% of the cluster mean. The test assertion previously
required that the max-min lease count delta was 0, when no outside
splits occurred; or 1 when the number of ranges was greater than the
number of stores.

The logging format is updated for easier debugging:

```
cpu outside bounds mean=23.9 tolerance=10.0% (±2.4) bounds=[21.5, 26.3]
   below  = []
   within = [s1: 22 (-6.7%), s2: 22 (-6.3%)]
   above  = [s3: 26 (+13.0%)]
...
cpu within bounds mean=25.7 tolerance=10.0% (±2.6) bounds=[23.2, 28.3]
   stores=[s1: 24 (-3.0%), s2: 24 (-2.9%), s3: 27 (+5.9%)]
```

resolves: https://github.com/cockroachdb/cockroach/issues/102801
resolves: https://github.com/cockroachdb/cockroach/issues/102823

Release note: None

kvoli committed:
roachtest: assert on CPU in rebalance/by-load

The `rebalance/by-load/replicas` roachtests periodically flake due to a
known limitation, where a cluster with a small number of ranges may not
be properly balanced due to heterogeneous localities (including
multi-store) #88829.

This commit updates the total number of ranges from 1 per-store, to 5
per-store for `rebalance/by-load/*` roachtests.

The roachtest asserted on the lease count, as a proxy for load, assuming
that the load evenly hits each lease for the created ranges. However,
the principal indicator of load in a read heavy workload is CPU.

This commit updates the test assertion to require that every store's CPU
is within 10% of the cluster mean. The test assertion previously
required that the max-min lease count delta was 0, when no outside
splits occurred; or 1 when the number of ranges was greater than the
number of stores.

The logging format is updated for easier debugging:

```
cpu outside bounds mean=23.9 tolerance=10.0% (±2.4) bounds=[21.5, 26.3]
   below  = []
   within = [s1: 22 (-6.7%), s2: 22 (-6.3%)]
   above  = [s3: 26 (+13.0%)]
...
cpu within bounds mean=25.7 tolerance=10.0% (±2.6) bounds=[23.2, 28.3]
   stores=[s1: 24 (-3.0%), s2: 24 (-2.9%), s3: 27 (+5.9%)]
```

resolves: #102801
resolves: #102823

Release note: None

kvoli committed:
roachtest: add sample period and source to ts_util

Previously, the utility functions for retrieving stats from the
timeseries database in roachtest wouldn't allow specifying a sample
period nor sources. This commit enhances ts_util.go to enable specifying
a sample period via `getMetricsWithSamplePeriod` and updates the
`tsQuery` to have a source field - which is applied on queries.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/102824)
<!-- Reviewable:end -->

aliher1911 commented on pkg/cmd/roachtest/tests/ts_util.go:
```diff
@@ -38,9 +38,13 @@ const (
 )
 
+// defaultSampleNanos is the default sampling period for getMetrics.
+const defaultSampleNanos = int64(time.Minute)
```
I think it would be more readable to have it as
```
const defaultSamplePeriod = time.Minute
```
and then use `defaultSamplePeriod.Nanoseconds()` only when using it in `TimeSeriesQueryRequest`.

aliher1911 commented on pkg/cmd/roachtest/tests/rebalance_load.go:
```diff
@@ -236,79 +268,110 @@ func registerRebalanceLoad(r registry.Registry) {
+// isLoadEvenlyDistributed checks whether the load for the stores given are
+// within tolerance of the mean. If the store loads are, true is returned as
+// well as reason, otherwise false. The function expects the loads to be
+// indexed to store IDs, see getLeaseCounts for example format.
```
> see `getLeaseCounts` for example format

I don't see getLeaseCounts anywhere in the code. Is it stale?

aliher1911 commented on pkg/cmd/roachtest/tests/rebalance_load.go:
```diff
@@ -125,23 +144,36 @@ func registerRebalanceLoad(r registry.Registry) {
 			}
 
-			return fmt.Errorf("timed out before leases were evenly spread")
+			return fmt.Errorf("timed out before CPU was evenly balanced")
```
I think it would be nicer for the reader to have error that would include reason. E.g. CPU is not evenly balanced after timeout: <load report>. Then it should be part of error message when test fails and GH issue is created.

aliher1911 reviewed (approved):
Looks good. Wouldn't it be still susceptible to some random variations in the cluster like backup running?

I put some minor nits.

kvoli committed:
roachtest: skip backup schedule on rebalance/*

The `rebalance/by-load` roachtests assert on the distribution of CPU
utilization between nodes in the cluster, having backups also running
could lead to unrelated flakes.

This commit updates cluster creation in `rebalance/by-load/` roachtests
to opt out of a backup schedule.

Epic: none

Release note: None

kvoli reviewed (commented):
It could still be susceptible. I pushed another commit to disable scheduled backups - I'd rather avoid any flakes on this test, since it has been high maintenance recently.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102824#-:-NV-jnHZCugC31cAgMWA:barnbj3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911, @smg260, and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/rebalance_load.go` line 176 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/102824#gh-1188535039:-NV-gxnK0Ovj0YdnQiAO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa9f4f0c6839bebe2e5feee6827ab1abe41320a3/pkg/cmd/roachtest/tests/rebalance_load.go#L176)):*
<details><summary><i>Previously, aliher1911 (Oleg) wrote…</i></summary><blockquote>

I think it would be nicer for the reader to have error that would include reason. E.g. CPU is not evenly balanced after timeout: <load report>. Then it should be part of error message when test fails and GH issue is created.
</blockquote></details>

Done.
___
*[`pkg/cmd/roachtest/tests/rebalance_load.go` line 322 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/102824#gh-1188534948:-NV-h4F-BVQh_-r1HTtp:b-hvtfcm) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa9f4f0c6839bebe2e5feee6827ab1abe41320a3/pkg/cmd/roachtest/tests/rebalance_load.go#L322)):*
<details><summary><i>Previously, aliher1911 (Oleg) wrote…</i></summary><blockquote>

> see `getLeaseCounts` for example format

I don't see getLeaseCounts anywhere in the code. Is it stale?
</blockquote></details>

Stale - updated to getStoreCPUFn.
___
*[`pkg/cmd/roachtest/tests/ts_util.go` line 42 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/102824#gh-1188534830:-NV-hWvGAyyqpXfip3c5:bl7bb6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa9f4f0c6839bebe2e5feee6827ab1abe41320a3/pkg/cmd/roachtest/tests/ts_util.go#L42)):*
<details><summary><i>Previously, aliher1911 (Oleg) wrote…</i></summary><blockquote>

I think it would be more readable to have it as
```
const defaultSamplePeriod = time.Minute
```
and then use `defaultSamplePeriod.Nanoseconds()` only when using it in `TimeSeriesQueryRequest`.

</blockquote></details>

Good idea, updated to this.


<!-- Sent from Reviewable.io -->

kvoli commented:
TYFTR

bors=aliher1911

kvoli commented:
bors r=aliher1911
